### ISSUE TITLE
Lightweight version of steps/shift_feats.sh

### ISSUE THREAD

**Issue by danpovey** [kaldi-asr/kaldi#1364]
@hhadian (and please ask @vimalmanohar for help as needed),
This is something that will be helpful in making the frame-shifted discriminative training (for chain and other LFR models) more efficient.  It will allow us to write the shifted versions of the features in a 'lightweight' way without dumping any real data.

[BTW, I looked into the relative size of the lattices and the features in the discriminative
egs, and they were about the same; so I decided that there is not very much to be gained
by writing out discriminative egs with all the differently-shifted lattices together in one eg; it's too much coding for too little gain in disk space, and we'd actually end up doing more I/O overall.]

Right now the example script egs/librispeech/s5/local/chain/run_tdnn_discriminative.sh shows how you can write out differently frame-shifted versions of the features using steps/shift_feats.sh.

I want a script called
utils/data/shift_feats.sh 
to be created with the following usage:
```
if [ $# != 3 ]; then
  echo " Usage: $0 <frame-shift> <srcdir> <destdir>"
  echo "e.g.: $0 -1 data/train data/train_fs-1"
  echo "The script creates a new data directory with the features modified"
  echo "using the program shift-feats with the specified frame-shift."
  echo "This program automatically adds the prefix "fs<frame-shift>-" to the"
  echo "utterance and speaker names.  See also utils/data/shift_and_combine_feats.sh"
fi
```
The script can be modified from limit_feature_dim.sh.
When calling copy_data.sh, you'll need the --utt-prefix and --spk-prefix options.

The program shift-feats.cc will need to be changed.  Currently it only supports table I/O,
and it doesn't have a conformant usage message.  It needs to support (r,w)xfilenames also.
The usage message should look more like the one from copy-feats.cc, e.g. like
the following:
```
        "Copy features, and possibly shift them while maintaining the num-frames.\n"
        "Usage: shift-feats [options] <feature-rspecifier> <feature-wspecifier>\n"
        "or:  shift-feats [options] <feats-rxfilename> <feats-wxfilename>\n"
        "e.g.: shift-feats --shift=-1 foo.scp bar.ark\n"
        "or: shift-feats --shift=1 foo.mat bar.mat\n"
        "See also: copy-feats, copy-matrix, select-feats, extract-rows, subset-feats,\n"
        "subsample-feats, splice-feats, paste-feats, concat-feats\n";
```
The script utils/data/shift_feats.sh will change the lines of feats.scp as follows, probably with
an inline perl or awk script:
if the line looks like
```
utt-123 abcdefg
```
it will be changed to the following (and
I'm ignoring issues of utterance prefixes, because copy_data.sh does that for you):
```
utt-123 shift-feats --print-args=false --shift=1 abcdefg - |
```
This is true generally, but if the rxfilename contains any pipe symbol then it needs to be quoted, e.g.:
```
utt-123 foo-bar abcdefg |[,123:456]
```
becomes:
```
utt-123 shift-feats --print-args=false --shift=1 'foo-bar abcdefg |[,123:456]' - |
```
and the script should die if the line of the feats.scp file already contained any single-quotes.
Make sure  that shift-feats does not print any log messages if invoked on (r,w)xfilenames.

You can also create a script`utils/data/shift_and_combine_feats.sh` with the following usage,
and to do this you can refer to `perturb_data_dir_speed_3way.sh`  [note: for the shift of zero you can
use the original input; there is no need to shift by zero.]
```
if [ $# != 3 ]; then
  echo "Usage: $0 <frame-subsampling-factor> <srcdir> <destdir>"
  echo "e.g.: $0 3 data/train data/train_fs3"
  echo "For use in perturbing data for discriminative training and alignment of"
  echo "frame-subsampled systems, this script uses utils/data/shift_feats.sh"
  echo "and utils/data_combine_data_dir.sh to shift the features"
  echo "<frame-subsampling-factor> different ways and combine them."
  echo "E.g. if <frame-subsampling-factor> is 3, this script will combine"
  echo "the data frame-shifted by -1, 0 and 1 (c.f. shift-feats)."
  exit 1
fi
```

It will be nice to have a standard script somewhere that can do exactly the same transformation on the ivector scp files; this will keep the example scripts concise.  You can propose something for this after studying egs/librispeech/s5/local/chain/run_tdnn_discriminative.sh and after doing the rest of this PR.

Please note: if the frame subsampling factor is `fs`, the range of shifts goes from `-(fs/2)` to `-(fs/2) + fs - 1`, it does not go from `-(fs/2)` to `(fs/2)` for even values of fs.

When this is done you can just remove steps/shift_feats.sh and replace the invocations with the new location.  This should all be done against the 'shortcut' branch because the invocations of the shift_data.sh script are in scripts that I've changed versus master, and I don't want merge conflicts.

Unlike limit_feature_dim.sh, you don't have to remove cmvn.scp (it will hardly change, so no need to regenerate).

**Comment by hhadian** [comment]
OK, will do it.

**Comment by danpovey** [comment]
resolved in #1477
